lib/mix/lib/mix/tasks/app.start.ex

```diff
@@ -160,6 +160,29 @@ defmodule Mix.Tasks.App.Start do
     end
   end
 
+  defp consult_app_file_maybe_in_archive(path) do
+    case :file.consult(path) do
+      {:ok, terms} ->
+        {:ok, terms}
+
+      {:error, :enotdir} ->
+        # The path could be located in .ez archive
+        case :erl_prim_loader.get_file(path) do
+          {:ok, bin, _full_name} ->
+            with {:ok, tokens, _} <- :erl_scan.string(String.to_charlist(bin)),
+                 {:ok, term} = :erl_parse.parse_term(tokens) do
+              {:ok, [term]}
+            end
+
+          :error ->
+            {:error, :enotdir}
+        end
+
+      {:error, reason} ->
+        {:error, reason}
+    end
+  end
```

Comment:
Here is perhaps a simpler implementation. Note we no longer wrap the term in a list though, so you will need to update the caller accordingly. :)

```
  defp consult_app_file_maybe_in_archive(path) do
    # The path could be located in .ez archive
    case :erl_prim_loader.get_file(path) do
      {:ok, bin, _full_name} ->
        with {:ok, tokens, _} <- :erl_scan.string(String.to_charlist(bin)) do
          :erl_parse.parse_term(tokens)
        end

      :error ->
        {:error, :enoent}
    end
  end
```